Fix TypedPredictor formatting with list output values

dbczumar opened the pull request:
Fixes #1567

dbczumar commented on dspy/adapters/chat_adapter.py:
```diff
@@ -2,16 +2,34 @@
+    info: FieldInfo
+
+
+# Built-in field indicating that a chat turn (i.e. a user or assistant reply to a chat
+# thread) has been completed.
+BuiltInCompletedOutputFieldInfo = FieldInfoWithName(name="completed", info=OutputField())
```
Because `_format_field_value` now requires information about:

1. Whether the field is an input or output field
2. The field's type annotation,

we cannot simply pass the string "completed" to `_format_field_value`. We need to pass a full `FieldInfo` object along with a field name. Accordingly, we define a class to represent this builtin field.

dbczumar commented on dspy/adapters/chat_adapter.py:
```diff
@@ -2,16 +2,34 @@
 field_header_pattern = re.compile(r"\[\[ ## (\w+) ## \]\]")
 
 
+class FieldInfoWithName(NamedTuple):
+    """
+    A tuple containing a field name and its corresponding FieldInfo object.
+    """
+
+    name: str
+    info: FieldInfo
```
Because `_format_field_value` now requires information about:

1. Whether the field is an input or output field
2. The field's type annotation,

we cannot simply pass string field names to `_format_field_value`. We need to pass a full `FieldInfo` object along with a field name. Accordingly, we define a class to represent both pieces of information.

mikeedjones replied:
does FieldInfo not have the field name/prefix as a member already?

dbczumar replied:
It doesn't :/

mikeedjones replied:
Ah well that would have been nice 😅

dbczumar commented on dspy/signatures/utils.py:
```diff
@@ -0,0 +1,10 @@
+from pydantic.fields import FieldInfo
+
+
+def get_dspy_field_type(field: FieldInfo) -> Literal["input", "output"]:
```
I noticed that we use `field.json_schema_extra.get("__dspy_field_type")` in many locations throughout DSPy to fetch the DSPy field type (input or output) from a pydantic field. To simplify this lookup going forward, I figured it would be useful to have a utility.

Happy to file a follow-up PR that replaces existing instances of `field.json_schema_extra.get("__dspy_field_type")` with a call to this utility.

mikeedjones replied:
Is an error always raised if there's no `dsyp_field_type`?

dbczumar replied:
Yeah, the current implementation attempts a dictionary access of the form `field.json_schema_extra["__dspy_field_type"]`, which fails with KeyError if the field is not present

mikeedjones replied:
ah sorry! - Think this check occurs when you add fields to the signature, `_SignatureMeta._validate_fields` so don't think it's necessary here? Can safely access by key after the signature is created.

Maybe slicker to see if a property can be added to `pydantic.Field` or if `pydantic.Field` can be subclassed.

dbczumar replied:
> Maybe slicker to see if a property can be added to pydantic.Field or if pydantic.Field can be subclassed.

I agree - does it make sense to pursue that as a followup to this PR, since it's a broader refactor?

> Think this check occurs when you add fields to the signature, _SignatureMeta._validate_fields so don't think it's necessary here

If this assumption ever fails for some case & we remove this logic, we'll raise a KeyError instead of a pretty error message. Happy to remove it, but not sure it hurts...

mikeedjones replied:
Yeah agree.

okhat commented:
Thank you so much @dbczumar ! This looks great to me. I realized I had made a mistake in discussing the if statement that decides between formatting options. I corrected my error in there and updated the comments to reflect this.

Here's the new snippet:

```python
if dspy_field_type == "input" or field_info.annotation is str:
    # If the field is an input field or has no special type reqs, format as a numbered list
    # so it's organized in a way suitable for presenting long context, i.e. not JSON.
    return format_input_list_field_value(value)
else:
    # If the field is an output field that has strict parsing reqs, format the value as
    # a stringified JSON Array. This ensures that downstream routines can parse the
    # field value correctly using methods from the `ujson` or `json` packages.
    return json.dumps(value)
```

okhat commented:
I would happily merge at this point but looks like we have plenty of failing tests. The tests failing are quite new (from #1595 cc @mikeedjones) so maybe we're learning something useful here about the changes that I missed in my review or about the tests.

dbczumar commented:
> Thank you so much @dbczumar ! This looks great to me. I realized I had made a mistake in discussing the if statement that decides between formatting options. I corrected my error in there and updated the comments to reflect this.
> 
> Here's the new snippet:
> 
> ```python
> if dspy_field_type == "input" or field_info.annotation is str:
>     # If the field is an input field or has no special type reqs, format as a numbered list
>     # so it's organized in a way suitable for presenting long context, i.e. not JSON.
>     return format_input_list_field_value(value)
> else:
>     # If the field is an output field that has strict parsing reqs, format the value as
>     # a stringified JSON Array. This ensures that downstream routines can parse the
>     # field value correctly using methods from the `ujson` or `json` packages.
>     return json.dumps(value)
> ```
> 
> tl;dr All inputs can safely be formatted as numbered lists, since parsing is not a concern. Outputs with no parsing requirements (i.e., outputs that are specified as strings) also fall under that, i.e. format as numbered lists. Otherwise, we use JSON.

Thanks @okhat ! I've incorporated your edit :)

dbczumar commented:
> I would happily merge at this point but looks like we have plenty of failing tests. The tests failing are quite new (from #1595 cc @mikeedjones) so maybe we're learning something useful here about the changes that I missed in my review or about the tests.

I think the test issues are caused by a bug in my change to `DummyLM`, which should hopefully be fixed in my latest couple of commits. Fingers crossed! :D 

Before merge, I'd like to introduce some test coverage to catch similar regressions in the future.

dbczumar commented on tests/functional/test_signature_typed.py:
```diff
@@ -1,9 +1,10 @@
-import pytest
 
 import pydantic
+import pytest
+
 import dspy
+from dspy.adapters.chat_adapter import _format_field_value
```
Just import adjustments made by the linter

dbczumar commented on tests/functional/test_signature_typed.py:
```diff
@@ -153,14 +160,18 @@ class MySignature(dspy.Signature):
 
 def test_dataclass():
-    from dataclasses import dataclass, asdict
+    from dataclasses import asdict, dataclass
```
Just import adjustments made by the linter

mikeedjones commented on dspy/utils/dummies.py:
```diff
@@ -170,6 +171,15 @@ def _use_example(self, messages):
 
     def __call__(self, prompt=None, messages=None, **kwargs):
+        def format_answer_fields(field_names_and_values: Dict[str, Any]):
+            print("FIELD NAMES AND VALUES", field_names_and_values)
```
can print statements be replaced with logs (or removed altogether?)

mikeedjones commented on dspy/utils/dummies.py:
```diff
@@ -170,6 +171,15 @@ def _use_example(self, messages):
                 return output["content"]
 
     def __call__(self, prompt=None, messages=None, **kwargs):
+        def format_answer_fields(field_names_and_values: Dict[str, Any]):
```
Massive nit, but `format_output_fields` to keep naming consistent with `OutputField`?

dbczumar replied:
Happy to make that adjustment if you feel strongly, but the constructor parameter to DummyLM that's ultimately being formatted here is called `answers`.

mikeedjones replied:
Ah you're right! I agree its better to keep local consistency now and change them both at some point?

dbczumar commented on tests/functional/test_functional.py:
```diff
@@ -452,6 +453,22 @@ class TestSignature(dspy.Signature):
     assert output == [0, 1, 2]
 
 
+def test_list_inputs_and_outputs():
```
@okhat I added a test case to hopefully prevent future regressions. On `main`, it fails during inference with:

```
        sections = [(k, "\n".join(v).strip()) for k, v in sections]

        fields = {}
        for k, v in sections:
            if (k not in fields) and (k in signature.output_fields):
                try:
                    fields[k] = parse_value(v, signature.output_fields[k].annotation) if _parse_values else v
                except Exception as e:
                    raise ValueError(
                        f"Error parsing field {k}: {e}.\n\n\t\tOn attempting to parse the value\n```\n{v}\n```"
                    )

        if fields.keys() != signature.output_fields.keys():
>           raise ValueError(f"Expected {signature.output_fields.keys()} but got {fields.keys()}")
E           ValueError: Expected dict_keys(['json_object']) but got dict_keys([])

../../dspy/adapters/chat_adapter.py:66: ValueError
```

mikeedjones commented on tests/functional/test_functional.py:
```diff
@@ -452,6 +453,22 @@ class TestSignature(dspy.Signature):
+    # for each element
+    expected_input_chat_content = "[[ ## input ## ]]\n[1] «3»\n[2] «4»\n[3] «5»"
+    input_chat_content = lm.get_convo(0)[0][1]["content"]
+    assert expected_input_chat_content in input_chat_content
```
From discussion here https://github.com/stanfordnlp/dspy/pull/1585#issuecomment-2395523404 think it makes sense to avoid checking the templated input/output in tests of modules/predictors?

mikeedjones replied:
I think we want tests _for_ `ChatAdaper` which check that all the stuff works, and then tests of the modules/predictors which are invariant under the exact inputs and outputs of the model? 

ATM `DummyLM` is built on the assumption that `ChatAdapter` is set as the adapter and will fail under another adapter.

mikeedjones commented on tests/functional/test_functional.py:
```diff
@@ -452,6 +453,22 @@ class TestSignature(dspy.Signature):
 
 
+def test_list_inputs_and_outputs():
+    lm = DummyLM([{"output": ["0", "1", "2"]}])
```
```suggestion
    lm = DummyLM([{"output": '["0", "1", "2"]'}])
```

I believe? - Language models produce strings!

dbczumar replied:
Great catch - fixed!
